src/hotspot/cpu/x86/downcallLinker_x86_64.cpp

```diff
@@ -133,7 +75,41 @@ RuntimeStub* DowncallLinker::make_downcall_stub(BasicType* signature,
   return stub;
 }
 
-void DowncallStubGenerator::generate() {
+static constexpr int RBP_BIAS = 16; // skip old rbp and return address
+
+void DowncallLinker::StubGenerator::pd_add_offset_to_oop(VMStorage reg_oop, VMStorage reg_offset,
+                                                         VMStorage tmp1, VMStorage tmp2) const {
+  if (reg_oop.is_reg()) {
+    assert(reg_oop.type() == StorageType::INTEGER, "expected");
+    if (reg_offset.is_reg()) {
+      assert(reg_offset.type() == StorageType::INTEGER, "expected");
+      __ addptr(as_Register(reg_oop), as_Register(reg_offset));
+    } else {
+      assert(reg_offset.is_stack(), "expected");
+      Address offset_addr(rbp, RBP_BIAS + reg_offset.offset());
+      __ addptr(as_Register(reg_oop), offset_addr);
+    }
+  } else {
+    assert(reg_oop.is_stack(), "expected");
+    assert(reg_offset.is_stack(), "expected");
+    Address offset_addr(rbp, RBP_BIAS + reg_offset.offset());
+    Address oop_addr(rbp, RBP_BIAS + reg_oop.offset());
+    __ movptr(as_Register(tmp1), offset_addr);
+    __ addptr(oop_addr, as_Register(tmp1));
+  }
+}
+
+static void runtime_call(MacroAssembler* _masm, address target) {
+  __ vzeroupper();
+  __ mov(r12, rsp); // remember sp
+  __ subptr(rsp, frame::arg_reg_save_area_bytes); // windows
+  __ andptr(rsp, -16); // align stack as required by ABI
+  __ call(RuntimeAddress(target));
+  __ mov(rsp, r12); // restore sp
+  __ reinit_heapbase();
+}
```

Comment:
This is a minor cleanup to share this code for the three use sites below.

Reply:
Question: `r12` does not need to remember?

According to [CallingSequences in OpenJDK Wiki](https://wiki.openjdk.org/display/HotSpot/CallingSequences), `r12` may be reserved for HeapBase if COOP is enabled.  
(`r12` is also used in another places in downcallLinker_x86_64.cpp without restoring...)

Reply:
You mean `reinit_heapbase` can be removed? I'm not sure whether the caller expects it to be preserved.

Note that `r12` is used in this case to save and restore `rsp`. This is needed since we access data in the frame relative to `rsp`.

Reply:
@YaSuenag `r12` is restored in `reinit_heapbase()` if needed and no, `r12` does not need remembering because it is a constant and can be restored from somewhere else.

Reply:
I think your code is fine. Restoring `r12_heapbase` at this point is not bad because `runtime_call` is only for slow paths. I don't think it should be moved.

Reply:
Thanks everyone! I understood `r12` is restored in `reinit_heapbase()`. It is necessary. I have no comments.